### PR TITLE
Odyssey Stats: Enable page view analytics

### LINKED ISSUE #77765: Odyssey Stats: Enable page view analytics
Leverage existing analytics architecture built into Calypso if possible.

### PR DESCRIPTION
**Pull request by jsnmoon**
Fixes #77765.

## Proposed Changes

* Enables the Analytics middleware for Odyssey Stats.
* Augments page view events for Jetpack Stats to include a boolean `is_odyssey` property.

## Privacy Implications

This enables page view tracking for the new Jetpack Stats experience (dubbed "Odyssey Stats" as it's Calypso-Stats-in-Jetpack).

This tracking is active only on wp-admin (specifically, `/wp-admin/admin.php?page=stats`) and is used to determine the reach and usage pattern for Odyssey Stats.

## Testing Instructions

**You may need to disable your adblocker for these analytics requests to fire.** 

#### On Calypso
* Open up this PR's live branch and navigate to Calypso Stats.
* Open your browser's network monitor and filter for requests including `calypso_page_view`.
* Ensure you see a request for `pixel.wp.com/t.gif`.
* Note that nothing should have changed for Calypso Stats. These steps for ensuring that everything works as expected.

#### On Odyssey
* Spin up a local Jetpack site using instructions from the Jetpack repository.
* Run `STATS_PACKAGE_PATH=/path/to/jetpack/projects/packages/stats-admin yarn dev` in `calypso/apps/odyssey-stats`.
* Navigate to Odyssey Stats at `/wp-admin/admin.php?page=stats` 
* Open your browser's network monitor and filter for requests including `calypso_page_view`.
* Ensure you see a request for `pixel.wp.com/t.gif` that has a defined `blog_id` and an `is_odyssey` query string set to **`true`**.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by jsnmoon** [comment]
@kangzj, I'm having trouble enabling the analytics middleware for Odyssey Stats, which I suspect is the reason why [PageViewTracker](https://github.com/Automattic/wp-calypso/blob/53e03079dc6e9547e2166c808bdea39c897bdd3a/client/my-sites/stats/site.jsx#L359) is not working as expected in Odyssey Stats.

If you have any pointers, it'd be greatly appreciated 🙏

**Comment by robfelty** [comment]
I haven't tested it myself, but I did search for tracks events and found some.

Does this also fire when Jetpack stats is loaded from the Jetpack dashboard? Should it? Or should we flag this some other way? I think it would be useful personally. We could get the info by filtering the Jetpack dashboard views to only people who have stats enabled, but I think that recording it directly would be more useful. If this also fires from the dashboard, we definitely want to record those two paths separately.

**Comment by kangzj** [comment]
HI @jsnmoon Thanks for tackling this and sorry for not being able to answer your question regarding adding the analytics middleware.

I works as described. However I have the same feeling as @robfelty - IMO we should fire a different event for Odyssey, as it's totally another product than Calypso - views from Odyssey shouldn't be counted as Calypso views and also we probably want to easily count Odyssey with another event.

Cheers

**Comment by jsnmoon** [comment]
Robert & Jasper: Your questions and suggestions are slightly different, so I'll address them separately.

> Does this also fire when Jetpack Stats is loaded from the Jetpack dashboard? (@robfelty)

No, this PR will only affect Odyssey Stats at `/wp-admin/admin.php?page=stats`. However...

> Should it? Or should we flag this some other way? I think it would be useful personally. We could get the info by filtering the Jetpack dashboard views to only people who have stats enabled, but I think that recording it directly would be more useful. (@robfelty)

If you think it'll be helpful, then it's probably worth implementing 🙂. We'll look into this in a future PR.

> we should fire a different event for Odyssey, as it's totally another product than Calypso - views from Odyssey shouldn't be counted as Calypso views and also we probably want to easily count Odyssey with another event. (@kangzj)

This is a valid concern. I'll investigate how to overhaul our overall approach for Calypso Stats and Odyssey Stats. Here's what I'm currently thinking for the scope of this PR:

- Preserve the existing `calypso_page_view` events on Calypso Stats pages.
- Add a new `calypso_stats_page_view` events to Calypso Stats pages.
- Add a new `odyssey_stats_page_view` events to Odyssey Stats pages.

What do you think?

**Comment by kangzj** [comment]
>    Preserve the existing calypso_page_view events on Calypso Stats pages.
    Add a new calypso_stats_page_view events to Calypso Stats pages.
    Add a new odyssey_stats_page_view events to Odyssey Stats pages.

Theoretically `calypso_stats_page_view` would not be necessary, but considering it would make the stats easier, I'm in line with the solution 👍

**Comment by robfelty** [comment]
> Preserve the existing calypso_page_view events on Calypso Stats pages.
> Add a new calypso_stats_page_view events to Calypso Stats pages.
> Add a new odyssey_stats_page_view events to Odyssey Stats pages.

Sounds like a good plan!

**Comment by robfelty** [comment]
> If you think it'll be helpful, then it's probably worth implementing 🙂. We'll look into this in a future PR.

Yeah - adding that in a future PR sounds good. It is somewhat tangential.

**Review by kangzj — APPROVED** [review]
I fixed up the Odyssey event [here](https://github.com/Automattic/wp-calypso/pull/77766/commits/93170e56b3c7bce0a5557bd4377479616815e919), where the event needs to be added to an exception list to be fired because it doesn't match the expected pattern.

And now it looks and tests well for both Calypso and Odyssey Stats 👍

**Comment by jsnmoon** [comment]
I'll merge and deploy this once Nosara #6324 has been merged.

**Comment by kangzj** [comment]
> I'll merge and deploy this once Nosara https://github.com/Automattic/wp-calypso/issues/6324 has been merged.

IMHO I guess we don't have to wait for that, as these are new events, which don't exist anyway.

File: client/my-sites/stats/stats-page-view-tracker/use-tracks.tsx
Comment (robfelty) [inline client/my-sites/stats/stats-page-view-tracker/use-tracks.tsx]: I should have thought of this earlier, but it didn't occur to me until I saw the other comments - how about we use `jetpack_odyssey_stats_page_view` instead. I don't think we want to add a whole new `source` for Tracks, unless we are planning on adding a bunch more `odyssey` Tracks events.

**Review by robfelty — DISMISSED** [review]

File: client/my-sites/stats/stats-page-view-tracker/use-tracks.tsx
Comment (jsnmoon) [reply client/my-sites/stats/stats-page-view-tracker/use-tracks.tsx]: All good. I waited for your review because I thought there might be something like this we missed 🙂

**Review by robfelty — APPROVED** [review]
Looks good now. Thanks for the update!

**Pull request merged**
